src/main/java/org/broadinstitute/hellbender/utils/haplotype/HaplotypeBAMWriter.java

```diff
@@ -0,0 +1,184 @@
+import org.broadinstitute.hellbender.utils.read.*;
+import org.broadinstitute.hellbender.utils.Utils;
+import org.broadinstitute.hellbender.utils.genotyper.ReadLikelihoods;
+
+import java.io.File;
+import java.util.Collection;
+import java.util.HashSet;
+import java.util.Set;
+
+/**
+ * A BAMWriter that aligns reads to haplotypes and emits their best alignments to a destination.
+ *
+ */
+public abstract class HaplotypeBAMWriter {
+    /**
+     * Allows us to write out unique names for our synthetic haplotype reads
+     */
+    private long uniqueNameCounter = 1;
+
+    private final static String READ_GROUP_ID = "ArtificialHaplotype";
+    private final static String HAPLOTYPE_SAMPLE_TAG = "HC";
+
+    protected final HaplotypeBAMDestination output;
+    private boolean writeHaplotypes = true;
+
+    /**
+     * Possible modes for writing haplotypes to BAMs
+     */
+    public enum Type {
+        /**
+         * A mode that's for method developers.  Writes out all of the possible
+         * haplotypes considered, as well as reads aligned to each
+         */
+        ALL_POSSIBLE_HAPLOTYPES,
+
+        /**
+         * A mode for users.  Writes out the reads aligned only to the called
+         * haplotypes.  Useful to understand why the caller is calling what it is
+         */
+        CALLED_HAPLOTYPES
+    }
+
+    /**
+     * Create a new HaplotypeBAMWriter of type "type" for writing SAMRecords to an output file
+     *
+     * @param type the type of the writer we want to create
+     * @param outputFile the destination, must not be null
+     * @param sourceHeader the header of the input BAMs used to make calls, must not be null.
+     * @return a new HaplotypeBAMWriter
+     */
+    public static HaplotypeBAMWriter create(final Type type, final File outputFile, final SAMFileHeader sourceHeader) {
+        if (type == null) {
+            throw new IllegalArgumentException("type cannot be null");
+        }
+
+        return create(type, new SAMFileDestination(outputFile, sourceHeader, READ_GROUP_ID, HAPLOTYPE_SAMPLE_TAG));
+    }
+
+    /**
+     * Create a new HaplotypeBAMWriter of type "type" for writing SAMRecords to and output destination
+     *
+     * Note that writer must have its presorted bit set to false, as reads
+     * may come in out of order during writing
+     *
+     * @param type the type of the writer we want to create
+     * @param destination the destination, must not be null
+     * @return a new HaplotypeBAMWriter
+     */
+    public static HaplotypeBAMWriter create(final Type type, final HaplotypeBAMDestination destination) {
+        if (destination == null) {
+            throw new IllegalArgumentException("writer cannot be null");
+        }
+        if (type == null) {
+            throw new IllegalArgumentException("type cannot be null");
+        }
+
+        switch ( type ) {
+            case ALL_POSSIBLE_HAPLOTYPES: return new AllHaplotypeBAMWriter(destination);
```

Comment:
move this coupling of enum values and subclasses to the enum class itself. Ie add a method `create(HaplotypeBAMDestination destination)` to the enum class and call the constructor passed at creation using a contstructor reference (eg AllHaplotypeBAMWriter::new). I dont like having the mapping of enum values to classes separate from the enum itself.